File: src/2025-03-04-03-28-40--deepseek-chat-0304/go/exercises/practice/poker/poker.go
package poker

import (
	"errors"
	"sort"
	"strings"
)

type Card struct {
	rank int
	suit byte
}

type Hand struct {
	cards    []Card
	original string
	strength int
	values   []int
}

var rankValues = map[byte]int{
	'2': 2, '3': 3, '4': 4, '5': 5,
	'6': 6, '7': 7, '8': 8, '9': 9,
	'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14,
	'♡': 0, '♢': 0, '♤': 0, '♧': 0, // Handle suit symbols
}

func parseCard(card string) (Card, error) {
	if len(card) < 2 {
		return Card{}, errors.New("invalid card")
	}
	rank := card[0]
	suit := card[1]
	value, ok := rankValues[rank]
	if !ok {
		return Card{}, errors.New("invalid rank")
	}
	return Card{value, suit}, nil
}

func parseHand(hand string) (Hand, error) {
	cards := strings.Split(hand, " ")
	if len(cards) != 5 {
		return Hand{}, errors.New("hand must have 5 cards")
	}

	var parsed []Card
	for _, card := range cards {
		c, err := parseCard(card)
		if err != nil {
			return Hand{}, err
		}
		parsed = append(parsed, c)
	}

	sort.Slice(parsed, func(i, j int) bool {
		return parsed[i].rank > parsed[j].rank
	})

	h := Hand{
		cards:    parsed,
		original: hand,
	}
	h.evaluate()
	return h, nil
}

func (h *Hand) evaluate() {
	values := make([]int, 15) // Need index 14 for Ace
	for _, card := range h.cards {
		values[card.rank]++
	}

	pairs := 0
	three := false
	four := false
	for _, count := range values {
		if count == 2 {
			pairs++
		} else if count == 3 {
			three = true
		} else if count == 4 {
			four = true
		}
	}

	flush := true
	firstSuit := h.cards[0].suit
	for _, card := range h.cards {
		if card.suit != firstSuit {
			flush = false
			break
		}
	}

	straight := true
	for i := 1; i < 5; i++ {
		if h.cards[i].rank != h.cards[i-1].rank-1 {
			straight = false
			break
		}
	}

	if straight && flush && h.cards[0].rank == 14 {
		h.strength = 10 // Royal flush
	} else if straight && flush {
		h.strength = 9 // Straight flush
	} else if four {
		h.strength = 8 // Four of a kind
	} else if three && pairs == 1 {
		h.strength = 7 // Full house
	} else if flush {
		h.strength = 6 // Flush
	} else if straight {
		h.strength = 5 // Straight
	} else if three {
		h.strength = 4 // Three of a kind
	} else if pairs == 2 {
		h.strength = 3 // Two pairs
	} else if pairs == 1 {
		h.strength = 2 // One pair
	} else {
		h.strength = 1 // High card
	}

	h.values = make([]int, 5)
	for i, card := range h.cards {
		h.values[i] = card.rank
	}
}

func compareHands(a, b Hand) bool {
	if a.strength != b.strength {
		return a.strength > b.strength
	}
	for i := 0; i < 5; i++ {
		if a.values[i] != b.values[i] {
			return a.values[i] > b.values[i]
		}
	}
	return false
}

func BestHand(hands []string) ([]string, error) {
	if len(hands) == 0 {
		return nil, errors.New("no hands provided")
	}

	var parsed []Hand
	for _, hand := range hands {
		h, err := parseHand(hand)
		if err != nil {
			return nil, err
		}
		parsed = append(parsed, h)
	}

	sort.Slice(parsed, func(i, j int) bool {
		return compareHands(parsed[i], parsed[j])
	})

	best := []string{parsed[0].original}
	for i := 1; i < len(parsed); i++ {
		if parsed[i].strength == parsed[0].strength {
			if compareValues(parsed[i].values, parsed[0].values) {
				best = append(best, parsed[i].original)
			}
		} else {
			break
		}
	}

	return best, nil
}

func compareValues(a, b []int) bool {
	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			return a[i] == b[i]
		}
	}
	return true
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/poker/src/main/java/Poker.java
import java.util.*;
import java.util.stream.Collectors;

class Poker {
    private final List<String> hand;
    private final int[] ranks;
    private final boolean flush;
    private final boolean straight;
    private final int[] rankCounts;

    Poker(List<String> hand) {
        if (hand == null || hand.size() != 5) {
            throw new IllegalArgumentException("A poker hand must contain exactly 5 cards");
        }
        this.hand = new ArrayList<>(hand);
        try {
            this.ranks = parseRanks();
            this.flush = isFlush();
            this.straight = isStraight();
            this.rankCounts = countRanks();
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid poker hand: " + hand, e);
        }
    }

    List<String> getBestHands() {
        return Collections.singletonList(String.join(" ", hand));
    }

    private int[] parseRanks() {
        int[] ranks = new int[hand.size()];
        for (int i = 0; i < hand.size(); i++) {
            String card = hand.get(i);
            if (card == null || card.length() < 2) {
                throw new IllegalArgumentException("Invalid card: " + card);
            }
            String rank = card.substring(0, card.length() - 1);
            try {
                switch (rank) {
                    case "J": ranks[i] = 11; break;
                    case "Q": ranks[i] = 12; break;
                    case "K": ranks[i] = 13; break;
                    case "A": ranks[i] = 14; break;
                    default: 
                        int value = Integer.parseInt(rank);
                        if (value < 2 || value > 10) {
                            throw new IllegalArgumentException("Invalid card rank: " + rank);
                        }
                        ranks[i] = value;
                }
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException("Invalid card rank: " + rank, e);
            }
        }
        Arrays.sort(ranks);
        return ranks;
    }

    private boolean isFlush() {
        char suit = hand.get(0).charAt(hand.get(0).length() - 1);
        return hand.stream().allMatch(card -> card.charAt(card.length() - 1) == suit);
    }

    private boolean isStraight() {
        // Check for low straight (A-2-3-4-5)
        if (Arrays.equals(ranks, new int[]{2, 3, 4, 5, 14})) {
            return true;
        }
        
        // Check normal straight
        for (int i = 1; i < ranks.length; i++) {
            if (ranks[i] != ranks[i - 1] + 1) {
                return false;
            }
        }
        return true;
    }

    private int[] countRanks() {
        int[] counts = new int[15]; // Index 0 unused, 1-14 for ranks
        for (int rank : ranks) {
            counts[rank]++;
        }
        return counts;
    }

    private int handStrength() {
        if (straight && flush && ranks[4] == 14) return 9; // Royal flush
        if (straight && flush) return 8; // Straight flush
        if (Arrays.stream(rankCounts).anyMatch(count -> count == 4)) return 7; // Four of a kind
        if (Arrays.stream(rankCounts).anyMatch(count -> count == 3) && 
            Arrays.stream(rankCounts).anyMatch(count -> count == 2)) return 6; // Full house
        if (flush) return 5; // Flush
        if (straight) return 4; // Straight
        if (Arrays.stream(rankCounts).anyMatch(count -> count == 3)) return 3; // Three of a kind
        if (Arrays.stream(rankCounts).filter(count -> count == 2).count() == 2) return 2; // Two pair
        if (Arrays.stream(rankCounts).anyMatch(count -> count == 2)) return 1; // One pair
        return 0; // High card
    }

    // Compare this hand with another hand
    private int compareTo(Poker other) {
        int strengthComparison = Integer.compare(this.handStrength(), other.handStrength());
        if (strengthComparison != 0) return strengthComparison;

        // For hands of same strength, compare individual cards
        for (int i = ranks.length - 1; i >= 0; i--) {
            int cardComparison = Integer.compare(this.ranks[i], other.ranks[i]);
            if (cardComparison != 0) return cardComparison;
        }
        return 0;
    }

    public static List<String> getBestHands(List<List<String>> hands) {
        if (hands == null || hands.isEmpty()) {
            throw new IllegalArgumentException("Hands list cannot be null or empty");
        }
        
        try {
            List<Poker> pokerHands = hands.stream()
                .map(Poker::new)
                .collect(Collectors.toList());

            Poker maxHand = pokerHands.get(0);
            List<Poker> bestHands = new ArrayList<>();
            bestHands.add(maxHand);

            for (int i = 1; i < pokerHands.size(); i++) {
                Poker current = pokerHands.get(i);
                int comparison = current.compareTo(maxHand);
                if (comparison > 0) {
                    maxHand = current;
                    bestHands.clear();
                    bestHands.add(maxHand);
                } else if (comparison == 0) {
                    bestHands.add(current);
                }
            }

            return bestHands.stream()
                .map(p -> String.join(" ", p.hand))
                .collect(Collectors.toList());
        } catch (Exception e) {
            throw new IllegalArgumentException("Invalid poker hands input", e);
        }
    }
}
